### ISSUE TITLE
systemd is remounting mounts the user just unmounted, or unmounting just mounted mounts

### ISSUE THREAD

**Issue by ghost** [systemd/systemd#11613]
**systemd version the issue has been seen with**
240-5

**Used distribution**
Debian unstable

**Expected behaviour you didn't see**
Systemd not mounting/unmounting random filesystems, except on boot, or if explicitly configured to do so. Remounting filesystems automatically does not make any sense in the normal case. No hard disks were hotplugged either.

**Unexpected behaviour you saw**
I umounted some filesystems with "umount". Then, a bit later, I noticed that systemd had mounted them again (they were mounted even though I didn't issue any mount commands, and there was text in journalctl that indicated systemd did it). In fact, I did not do anything other than running fdisk after the umount command. I noticed only because fdisk complained that it couldn't update the kernel's partition tables.

There was some delay until it remounted. This is how I could use successfully fdisk to issue the ioctl to update the kernel partition tables, and I didn't investigate further.

I also observed that systemd _unmounted_ file systems I mounted via "mount". I was plenty of confused that the mount command succeeded (and no errors in the kernel log), but the mount was not present. Once I ran journalctrl, it was clear that journalctl unmounted them as soon as the mounts appeared. I don't remember if the devices/disk UUIDs or the mount target paths were in fstab or any of the apparently auto generated units.

Unlike the remounts, the unmounts were instant.

**Steps to reproduce the problem**
Unknown. It might be worth mentioning that I'm using /etc/fstab to configure the system's filesystems.

Unfortunately I don't have any logs anymore. journalctl says: "Data from the specified boot (-1) is not available: No such boot ID in journal". Also I'm finished reshuffling my disks now.

I don't even know what causes this. systemd was installed only recently, and I didn't configure anything systemd related other than logind and networkd. (Also tried to convince journald to put log files on the disk, though it still didn't preserve them across boots?)

**Comment by wm4** [comment]
It looks like only the unmount-on-mount issue is potentially a duplicate. Since the logs are (apparently) lost, I can't really argue against it. It looks vaguely similar and it's completely possible that it's exactly the same.

I don't see the automatic remount problem in that issue, though.

**Comment by bl33pbl0p** [comment]
If you saw the "mount manually stuff in fstab, have systemd unmount it immediately" in emergency mode, that is #7101. It happens when udev isn't running (so all device units are dead, and since mounts are bound to devices, systemd will immediately unmount it in the device's absence). As you note correctly, the unmounting is instant.

I am not sure about the first one, and without any logs it is hard to guess.

**Comment by bl33pbl0p** [comment]
Also, jobs aren't interruptible, you can pass systemd.debug_shell on the kernel command line which spawns a root shell on tty9, which you can then switch to and invoke systemctl list-jobs, see the job ID of the running job, and then systemctl cancel JOB_ID.

**Comment by tomty89** [comment]
Maybe also a duplicate of #11263 / #11373.

**Comment by keszybz** [comment]
So this is not a duplicate of any of the other issues, based on my understanding.
We have two mechanisms at play here:
1. .device units had an implicit Wants dependency on the corresponding .mount unit.  This meant that the device unit appearing would always cause the mount to happen. This has now been removed by 142b8142d7bb84f07ac33fc00527a4d48ac8ef9f.
2. The .mount unit is wanted by other things, most notably local-fs.target. When any unit is started that has a transitive dependency on local-fs.target, the .mount unit will also be started, if it wasn't started already. Since pretty much anything depends on local-fs.target, chances are that the .mount unit will be restarted very quickly, unless it is removed from the dependency tree. To remove it, it'd be necessary to e.g. mask the entry in /etc/fstab and do systemctl daemon-reload. Doable, but annoying.

**Comment by arvidjaar** [comment]
> When any unit is started that has a transitive dependency on local-fs.target, the .mount unit will also be started, if it wasn't started already.

Similar issues came up before. It is still not clear whether this is conscious decision or just side effect of implementation. It is not how I interpret documentation in any case. `If this unit gets activated, the units listed here will be activated as well`, but **this** unit (`local-fs.target`) is not being activated because it is already active.

**Comment by bl33pbl0p** [comment]
@keszybz

Yes, but the bug I pointed to was about systemd forcibly \*unmounting\* the mount as soon as you do it (if a mount unit was generated for it, and it becomes active, say because it was in fstab) in absence of device units (so when udev isn't running which is emergency mode) because the generated dependency is BindsTo=.

Anyway, the core problem here is this: the transaction computer will not take into account unit state.

This means if you request startup of a unit, the job set is generated statically, following loaded dependencies (so if I have Wants=local-fs.target, for instance, it will also process JOB_START for local-fs.target all over again). This means if some deps of such a unit are down, they will get a start job enqueued for them. For others, systemd relies on the property that requested and target state match so no action will be invoked on the unit, becoming a NOP.

Now, this is a deliberate design decision, afaics, and it has been a source of constant pain all over these years...

Consider:

https://github.com/systemd/systemd/issues/11440

The bug describes the entire scenario well, and there are two key problems: systemd computes and propagates jobs on units
that already need no state transition (i.e. stop on something already stopped, and so on), and systemd is not enable to
resolve conflicting job types inside the same transaction on the same unit (unlike conflicting jobs across transactions
where they can replace the installed one).

The first one also means when you stop something already stopped, it will follow things like PartOf= and BindsTo=, I am sure this is surprising. People have often been unable to debug problems stemming from this kludge, see https://github.com/systemd/systemd/issues/2830.

I have yet to hear a convincing case where this behaviour is useful (I know there are a few) BUT whether it warrants being the default, and not some job mode (recursive-static etc) that systemd can apply internally.

**Comment by bl33pbl0p** [comment]
Now, onto an idea for a solution.

We should exploit something systemd already does, for restart jobs. It forwards a JOB_TRY_RESTART when you enqueue a JOB_RESTART on a unit.

Understanding why it is the way it is gives one a helpful intuition: you usually restart things after the system has reached an **intended** state (i.e. in the most common case of restarts, we are in a running system, where the services are already up). This means we must have pulled in weak dependencies once during the initial transaction, and redoing that if they went down is not needed (otherwise the user must have used a stronger dependency type that does a state change to us too). So we map JOB_RESTART to JOB_TRY_RESTART for all dependencies, which **collapses** to either JOB_RESTART or JOB_NOP before appearing in the transaction.

I think this is how things should be in general in every transaction except the initial one, for units that are already active. We can have exceptions by virtue of job modes, but experience shows things need to be changed. I also think you can reasonably pull it off without breaking things, with some minimal changes to transaction.c.

(As a side note, having a way to specify the job mode like Wants=foo.service/static would also be helpful for those who wanted the old behaviour).
